### ISSUE TITLE
Specify token lifetime for aws_eks_cluster_auth

### ISSUE THREAD

**Issue by innovate-invent** [hashicorp/terraform-provider-aws#16415]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Long running deployments will time out after 20min with `Error: Unauthorized`. Please allow specifying the token lifetime in aws_eks_cluster_auth to work around this issue.

### New or Affected Resource(s)

* aws_eks_cluster_auth

### Potential Terraform Configuration

```hcl
data "aws_eks_cluster" "cluster" {
  name = local.cloud.eks.cluster_id
}

data "aws_eks_cluster_auth" "cluster" {
  name = local.cloud.eks.cluster_id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.cluster.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
  token                  = data.aws_eks_cluster_auth.cluster.token
  load_config_file       = false
}
```

**Comment by bflad** [comment]
Hi @innovate-invent 👋  Thank you for raising this.

To be upfront, I'm not sure if we can support this enhancement. The token generation code, originally from https://github.com/kubernetes-sigs/aws-iam-authenticator/ and now hard copied into this codebase at `aws/internal/service/eks/token` has some comments that seem to indicate that STS may have some hardcoded 15 minute expirations.

Given the long deployment times in your environment, you may be best to use a different Kubernetes authentication method beyond this data source since it is fetched once and done. There are likely options out there supported by the Terraform Kubernetes Provider that can dynamically refresh the authentication as necessary.

**Comment by innovate-invent** [comment]
Just looking at this documentation, I don't think there is any limitation on the API endpoint: https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html#API_GetSessionToken_RequestParameters

**Comment by camdentech-ts** [comment]
Great idea! I originally separated my kubernetes changesets from everything else in Terraform to get around this, but in the case of queued builds in terraform cloud (as is often the case when I merge dev changes into master) or if I don't immediately approve a plan then the token will be expired at the point an apply is attempted

**Comment by juozasget** [comment]
I will try to have a look at this and get back with some results.

**Comment by jw-maynard** [comment]
@innovate-invent It looks like the note in the code that @bflad referenced is correct. Expiration longer than 15 minutes is reserved solely for use with S3. All other services that support pre signed urls (like EKS) have a hard coded 15 minute expiration in STS. The pertinent discussion in the aws go SDK project can be found here https://github.com/aws/aws-sdk-go/issues/2167 It is dated I guess but as it's a security thing it seems like something AWS is unlikely to have changed their minds about.

**Comment by innovate-invent** [comment]
That is unfortunate, can the token be used to renew itself?
Perhaps the provider can try and renew its token.

**Comment by jw-maynard** [comment]
I was trying to think of a theoretical way to do this but I don't think it's possible. Regenerating the token would logically have to take place in the k8s provider, but that provider is supposed to be cloud provider agnostic. Recreating this token would require that provider to do AWS specific stuff to refresh the token I think. :( 

This EKS module (https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest?tab=outputs) uses a null_resource that waits on the cluster to be ready and then makes the data sources wait on the null_resource. That seems to solve a lot of these issues but not all of them. When I've run into these Unauthorized errors though usually just a re-run of a plan/apply takes care of the problem.

That said I've run into a new issue where I can't change things about the cluster (like upgrading the version) by just changing my TF and plan/applying. For some reason the authentication completely falls apart and the k8s provider starts trying to talk to `localhost` from some unknown reason.

**Comment by innovate-invent** [comment]
I guess there is no solution.
My deployments time out not because they are waiting for the k8s instance to come available, but because there are potentially long running k8s jobs that I have to wait on to continue my deployment. I'll just add more `-target` calls to the bodgy bash script I am already having to maintain to get around terraforms limitations.
